### PR TITLE
GopherCAN 2024 Cleanup

### LINKED ISSUE #116: Attach callbacks to 11-bit ID messages
Currently, functions can be triggered over CAN by calling `add_custom_can_func` with a function pointer and sending a 29-bit ID message with the command ID in the data field.

It would be useful to be able to attach callbacks to messages with a specified 11-bit ID. For example, there's a [special case](https://github.com/gopher-motorsports/gophercan-lib/blob/adf9a96476790acc05a346b822e8d98113827372/GopherCAN.c#L843) for responding to the lap beacon parameter. 

Callbacks would allow 11-bit ID messages to trigger functions on other modules. These triggers could be configured like any other parameter in the .yaml, removing the need for EXT_ID messages and special cases like above.

### PR DESCRIPTION
**Pull request by j0la**
The goal of these changes is to:
- Remove/refactor old/unused/deprecated functionality
- Minimize library size by merging files where appropriate and removing bloat
- Make it easier to read through the code and understand what GopherCAN does and how to use it
- Properly document how GopherCAN works and how to use it in the README

Overall, three files have been merged into `GopherCAN.c` and it's still ~100 lines shorter with all of the functionality from before, new features, a simpler API (imo), and better documentation.

Resolves #116 

## Notable Changes
- Created a `BUS_ID` enum from YAML
    - YAML bus definitions were previously unused, this change enforces consistency
- Removed `bx_type` and `module_id` parameters from `init_can`
    - CAN2 is a slave bxCAN in dual CAN configuration [RM0390 30.2](https://www.st.com/resource/en/reference_manual/rm0390-stm32f446xx-advanced-armbased-32bit-mcus-stmicroelectronics.pdf#%5B%7B%22num%22%3A1643%2C%22gen%22%3A0%7D%2C%7B%22name%22%3A%22XYZ%22%7D%2C67%2C755%2Cnull%5D). This doesn't need to be a parameter
    - Module ID should now be defined in `GopherCAN_config.h` e.g. `#define THIS_MODULE_ID 1`. This prevents mismatches from multiple calls to `init_can`.
    - The purpose of `init_can` is to tie a `BUS_ID` to an hcan instance and start the peripheral
- CAN message buffers are now generated in `GopherCAN_network`
    - This allows the number/names of buffers to follow the buses defined in a YAML
- Created `CAN_BUS` and `CAN_BUFFER` structs
    - Generated based on bus names in YAML
    - Removes the need for `NUM_OF_BUSSES` define
    - Should make it much easier to get the CAN instance, mutex, buffers, etc associated with a `BUS_ID`
    - Will also make it easier to specify which bus to send a message on if people want that feature
- Removed `GopherCAN_names`
    - Breaks compatibility with the old C data parser. This will happen in 2024 anyway because of a new data logging format.
- Reworked message filters. Each filter set can be toggled in `GopherCAN_config`.
- Reworked CAN routing with the new `CAN_BUS` struct.
- Created `__weak` event handlers
    - These allow attaching some custom functionality to events e.g. message received, error message
    - `GCAN_RxCallback()` is now called `GCAN_onRX` following this format, called after `service_can_rx_hardware()`
    - Created `GCAN_onError()`
- Added the ability to attach callbacks to parameter groups (`attach_callback_std()`)
    - When the specified group is received, the attached function pointer will be invoked
    - Removed the lap beacon detection since this can now be implemented as a callback on the appropriate module
- Reworked commands to follow the callback style (`attach_callback_cmd()`)
    - Removed the ability to enable/disable command functions, `do_nothing()`, and `mod_custom_can_func_state()` -- all of this functionality can be replaced by just not attaching a callback, or attaching `NULL`
- Simplified command and error definition in YAML configuration
- `send_parameter()` now takes a `GCAN_PARAM_ID`, no longer need to cast to a `CAN_INFO_STRUCT`
- Proper README documentation

## Example simplifications from these changes
Old `init_can`
```c
// attach hcan and mutex to appropriate buffer
#if NUM_OF_BUSSES > 2
    if (bus_id == GCAN2) {
        txbuff2.hcan = hcan;
        osMutexDef(txbuff2_mutex);
        txbuff2.mutex = osMutexCreate(osMutex(txbuff2_mutex));
    }
#endif
#if NUM_OF_BUSSES > 1
    if (bus_id == GCAN1) {
        txbuff1.hcan = hcan;
        osMutexDef(txbuff1_mutex);
        txbuff1.mutex = osMutexCreate(osMutex(txbuff1_mutex));
    }
#endif
    if (bus_id == GCAN0) {
        txbuff0.hcan = hcan;
        osMutexDef(txbuff0_mutex);
        txbuff0.mutex = osMutexCreate(osMutex(txbuff0_mutex));
    }
```

New `init_can`
```c
// attach hcan to designated bus
CAN_BUS* bus = BUSES[bus_id];
bus->hcan = hcan;
```

Old CAN routing
```c
#if defined(CAN_ROUTER)
static void rout_can_message(CAN_HandleTypeDef* hcan, CAN_MSG* message)
{
	MODULE_ID dest_module = GET_ID_DEST(message->header.ExtId);

	// TX buffer on the CAN bus this message was received on
	CAN_MSG_RING_BUFFER* orig_tx_buffer = choose_tx_buffer_from_hcan(hcan);

	// TX buffer the message is intended for
	CAN_MSG_RING_BUFFER* dest_tx_buffer = choose_tx_buffer_from_dest_module(dest_module);

	// message is intended for all modules, retransmit on a new bus
	if (dest_module == ALL_MODULES_ID) {
#if NUM_OF_BUSSES > 2
		if (&txbuff2 != orig_tx_buffer) add_message_by_highest_prio(&txbuff2, message);
#endif
#if NUM_OF_BUSSES > 1
		if (&txbuff1 != orig_tx_buffer) add_message_by_highest_prio(&txbuff1, message);
#endif
		if (&txbuff0 != orig_tx_buffer) add_message_by_highest_prio(&txbuff0, message);
		return;
	}

	// message is intended for this module, don't retransmit
	if (dest_module == this_module_id) return;

	// message was already on the correct bus, don't retransmit
	if (dest_tx_buffer == orig_tx_buffer) return;

	// remove message from RX buffer and retransmit
	remove_from_front(&rxbuff);
	add_message_by_highest_prio(dest_tx_buffer, message);
}
#endif // #if defined(CAN_ROUTER)
```

New CAN routing
```c
#ifdef ENABLE_ROUTING
  MODULE_ID dest_module = GET_ID_DEST(message->header.ExtId);
  BUS_ID dest_bus = module_bus_number[dest_module];
  
  if (dest_bus != get_bus_from_hcan(hcan)) {
	  // try to retransmit message on the correct bus
	  if (dest_module == ALL_MODULES_ID) {
		  for (U8 i = 0; i < ALL_BUSSES; i++) {
			  CAN_BUS *bus = BUSES[i];
			  add_message_by_highest_prio(bus, message);
		  }
	  } else if (dest_module != THIS_MODULE_ID) {
		  CAN_BUS *bus = BUSES[dest_bus];
		  add_message_by_highest_prio(bus, message);
	  }
  }
#endif
```

## Attaching Callbacks Example
```c
void demo() {
    ...
}

void setup() {
    init_can(&hcan1, GCAN0);
    attach_callback_std(u8_tester.info.GROUP_ID, demo);
}
```

### CODE REVIEW

File: network_autogen/configs/dyno-can-config.yaml
Comment (AlexT2052) [inline network_autogen/configs/dyno-can-config.yaml:1]: Why was this deleted? Did I miss a place it was moved? This isn't something that will be re-made for this year.

Comment (j0la) [reply network_autogen/configs/dyno-can-config.yaml:1]: It's temporary, they can be copy/pasted and committed again if this gets merged. I took them out since a few of the YAML fields were simplified and I didn't want to update them all right away.
